feat: show kubernetes context details in settings

cdrage opened the pull request:
feat: show kubernetes context details in settings

### What does this PR do?

* The first implementation for the Kubernetes context screen
* Shows Kubernetes context details (name, cluster, server URL)

In future PR's:
* Ways to determine the name / assign correct icon's to the provider
* Reactively update details (amount of pods, deployments, etc.)

### Screenshot/screencast of this PR

<!-- Please include a screenshot or a screencast
explaining what is doing this PR -->

![Screenshot 2023-10-30 at 7 04 07 PM](https://github.com/containers/podman-desktop/assets/6422176/b3a30a36-2385-4eb6-8b12-d937fc14f866)



### What issues does this PR fix or reference?

<!-- Include any related issues from Podman Desktop
repository (or from another issue tracker). -->

Closes: https://github.com/containers/podman-desktop/issues/3919


Completes first phase of https://github.com/containers/podman-desktop/issues/4416

### How to test this PR?

* Have a ~/.kube/config
* Click on Kubernetes in the settings

<!-- Please explain steps to reproduce -->

cdrage commented:
UI / implementation notes @mairin @benoitf 

UI:
* Implemented the design based upon Providers / CLI Tools section
* Will improve as we complete other parts of the epic: https://github.com/containers/podman-desktop/issues/4416

Implementation:
* Implemented all that can be done from the config file. The rest of the information (online status, amount of pods, deployments, etc.) needs to be determined in the background / "loading" as some kubernetes clusters take a while to query, etc. This will be done in: 

Follow up issues after this PR:
- https://github.com/containers/podman-desktop/issues/4561
- https://github.com/containers/podman-desktop/issues/4562

deboer-tim commented:
Looking good 👍🏼, will do full review tomorrow. The only things that are standing out quickly are:
- The charcoal-900 on cluster/server/user looks like black, wondering if it should be backed off to be closer to design.
- No vertical line in the design? Curious why this was added.

slemeur commented:
@mairin @cdrage : In my experience the names given to the kubecontext are always quite long. I'm a bit worry about how it will scale if the name is awfully long.

Here, with one context created by connecting to DevSandbox:
<img width="738" alt="Screenshot 2023-10-31 at 14 26 09" src="https://github.com/containers/podman-desktop/assets/1636769/34fbd414-1692-4f54-bcf6-153f69c31dcb">

I'm wondering if the name of the context shouldn't be better as a title of the card, so it would benefit from the full width card, be more readable than split on multiple lines? If it requires more space than available, then we could a text-overflow and a tooltip?

deboer-tim commented:
Yes, +1 to making it a full-width title please. :)

<img width="669" alt="Screenshot 2023-10-31 at 10 04 13 AM" src="https://github.com/containers/podman-desktop/assets/19958075/22dcfbec-a6e4-4e92-98d5-2e911ad1bbf5">

cdrage commented:
> Yes, +1 to making it a full-width title please. :)
> 
> <img alt="Screenshot 2023-10-31 at 10 04 13 AM" width="669" src="https://user-images.githubusercontent.com/19958075/279401432-22dcfbec-a6e4-4e92-98d5-2e911ad1bbf5.png">

I'll make that change! Thanks for pointing it out. Got to love long cluster names haha.

mairin commented:
![Screenshot from 2023-10-31 06-46-24](https://github.com/containers/podman-desktop/assets/799683/448681c3-05a1-4beb-b6ab-26314ca5edd6)
![Screenshot from 2023-10-31 06-47-37](https://github.com/containers/podman-desktop/assets/799683/63ea886a-5897-4c5e-a9e4-d4ddf61ba1dd)

I love seeing this become real :) I noticed some scaling issues on my screens (this is on Fedora 38.)

deboer-tim commented:
A few more concerns:
- What is the current context and how do I change it? I found a mockup with tiles that had a blue left border, not sure if it should be the same with these rows (and if it should be purple).
- The server link should be purple (maybe link button?), not blue.
- I could only see 3ish of my current 5 contexts on a small window, and the last screen cap above only has 2. I think we need a full-width title, but that might make the vertical space worse in some cases. I think we need to keep an eye on vertical space so more fit on screen at a time, e.g. reduce the top/bottom padding of those values and maybe use ellipsis with tooltips so long cluster/server/user values don't go two lines.

cdrage commented:
> A few more concerns:
> 
> * What is the current context and how do I change it? I found a mockup with tiles that had a blue left border, not sure if it should be the same with these rows (and if it should be purple).

Will be added in https://github.com/containers/podman-desktop/issues/4417 this PR is only implementing the page to display the context (for now).

> * The server link should be purple (maybe link button?), not blue.

Updated! Thanks for pointing that out, I've actually opened up https://github.com/containers/podman-desktop/pull/4577 when implementing this as well. So if someone is testing this PR, you'll notice a very subtle unaligned link, which will be fixed in https://github.com/containers/podman-desktop/pull/4577

> * I could only see 3ish of my current 5 contexts on a small window, and the last screen cap above only has 2. I think we need a full-width title, but that might make the vertical space worse in some cases. I think we need to keep an eye on vertical space so more fit on screen at a time, e.g. reduce the top/bottom padding of those values and maybe use ellipsis with tooltips so long cluster/server/user values don't go two lines.

Agreed! Added the full-width title now. 

![Screenshot 2023-10-31 at 5 19 09 PM](https://github.com/containers/podman-desktop/assets/6422176/c44de598-7cf8-4a10-b8f7-72a6b52adad9)

Ready for another review 👍

lstocchi reviewed (commented):
LGTM, just a tiny issue with the hover background that is a bit smaller with long names

![hover-background](https://github.com/containers/podman-desktop/assets/49404737/f4c53f53-c0d8-4d18-b3cb-6566d3d62136)

cdrage commented:
> LGTM, just a tiny issue with the hover background that is a bit smaller with long names
> 
> ![hover-background](https://user-images.githubusercontent.com/49404737/279945135-f4c53f53-c0d8-4d18-b3cb-6566d3d62136.gif) [ ![hover-background](https://user-images.githubusercontent.com/49404737/279945135-f4c53f53-c0d8-4d18-b3cb-6566d3d62136.gif) ](https://user-images.githubusercontent.com/49404737/279945135-f4c53f53-c0d8-4d18-b3cb-6566d3d62136.gif) [ ](https://user-images.githubusercontent.com/49404737/279945135-f4c53f53-c0d8-4d18-b3cb-6566d3d62136.gif)

I'll implement this, looks like we have some odd text wrapping going on there too.

cdrage commented:
@deboer-tim Updated! I had the columns set incorrectly and `ml-3` added in the wrong place. It now shows the full hover.

https://github.com/containers/podman-desktop/assets/6422176/e045cfff-2f80-4c5e-8313-459dcd1a8ac7

deboer-tim reviewed (approved):
LGTM, worked well, and scaled correctly.

lstocchi reviewed (approved):
LGTM, thanks Charlie!! 🚀

benoitf commented on packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte:
```diff
@@ -0,0 +1,75 @@
+
+$: contexts;
+
+onMount(async () => {
+  // Retrieve both the contexts and clusters
+  let k8sContexts = await window.kubernetesGetContexts();
+  let k8sClusters = await window.kubernetesGetClusters();
+
+  // Convert them to KubeContextUI so we can safely render them.
+  contexts = getKubeUIContexts(k8sContexts, k8sClusters);
+});
+</script>
```
From what I see we collect only the contexts/clusters when we display the page

so if I create a cluster, delete, etc while being on the page it won't be refreshed.

I think we need to use the store pattern and events to correctly have the UI being refreshed from changes and not from when we display the UI

cdrage replied:
Yes, we will be adding this as part of the second phase since we need to add actions such as delete / create which would be in a store: https://github.com/containers/podman-desktop/issues/4416

benoitf commented on packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte:
```diff
@@ -0,0 +1,75 @@
+      aria-label="no-resource-panel"
+      icon="{EngineIcon}"
+      title="No Kubernetes contexts found"
+      message="Check that ~/.local/config exists or KUBECONFIG environment variable has been set correctly."
```
IMHO it's `$HOME/.kube/config` not .local

We also have a property in Podman Desktop so probably ask user to check the value of this property as well (KUBECONFIG won't work AFAIK)

cdrage replied:
Updated! Good catch, it's suppose to be .kube/ not .local/

cdrage commented:
Pushed the changes to the text output. For the responsive / store, we'll implement that when we add the create / delete functionality as we will need to have that within a store / responsive to changes.

benoitf reviewed (approved):
approving as @cdrage said he'll do the store implementation in the next step
